sql/src/main/java/io/crate/planner/consumer/GroupingSubselectConsumer.java

```diff
@@ -0,0 +1,292 @@
+
+    private static class Visitor extends RelationPlanningVisitor {
+
+        private final ProjectionBuilder projectionBuilder;
+
+        public Visitor(ProjectionBuilder projectionBuilder) {
+            this.projectionBuilder = projectionBuilder;
+        }
+
+        @Override
+        public Plan visitQueriedSelectRelation(QueriedSelectRelation relation, ConsumerContext context) {
+            QuerySpec querySpec = relation.querySpec();
+
+            // Check if group by is present - if not skip and continue with next consumer in chain.
+            if (!querySpec.groupBy().isPresent()) {
+                return null;
+            }
+
+            Plan plan = context.plannerContext().planSubRelation(relation.subRelation(), context);
+
+            addWhereFilterProjectionIfNecessary(plan, querySpec, relation);
+
+            return createPlan(
+                plan,
+                context,
+                SplitPoints.create(querySpec),
+                querySpec.groupBy().get(),
+                querySpec.outputs(),
+                querySpec,
+                projectionBuilder
+            );
+        }
+    }
+
+    public static Plan createPlan(Plan plan,
+                                  ConsumerContext context,
+                                  SplitPoints splitPoints,
+                                  List<Symbol> groupKeys,
+                                  List<Symbol> outputs,
+                                  QuerySpec querySpec,
+                                  ProjectionBuilder projectionBuilder) {
+        if (isExecutedOnHandler(context, plan)) {
+            addNonDistributedGroupProjection(plan, splitPoints, groupKeys, projectionBuilder);
+            addFilterProjection(plan, splitPoints, groupKeys, querySpec);
+            addTopN(plan, splitPoints, groupKeys, querySpec, context, outputs);
+        } else {
+            addDistributedGroupProjection(plan, splitPoints, groupKeys, projectionBuilder);
+            plan = createReduceMerge(plan, splitPoints, groupKeys, querySpec, context, outputs, projectionBuilder);
+        }
+
+        return plan;
+    }
+
+    /**
+     * Adds the where filter projection if necessary.
+     */
+    private static void addWhereFilterProjectionIfNecessary(Plan plan,
+                                                            QuerySpec querySpec,
+                                                            QueriedSelectRelation relation) {
+        WhereClause where = querySpec.where();
+        if (where.hasQuery() || where.noMatch()) {
+            FilterProjection filterProjection = ProjectionBuilder.filterProjection(
+                relation.subRelation().fields(),
+                where
+            );
+            plan.addProjection(filterProjection, null, null, null);
+        }
+    }
+
+    /**
+     * Returns true for non distributed and false for distributed execution.
+     */
+    private static boolean isExecutedOnHandler(ConsumerContext context, Plan plan) {
```

Comment:
is this abstraction method worth it? Has the same number of arguments and the args are even "bigger" in that pull in classes that are harder to create